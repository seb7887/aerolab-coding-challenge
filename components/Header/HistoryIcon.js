import React from 'react';

const HistoryIcon = () => (
  <img src='/static/images/history.svg' alt='History' />
);

export default HistoryIcon;
