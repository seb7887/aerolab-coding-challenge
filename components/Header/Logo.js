import React from 'react';

const Logo = () => (
  <img src='/static/images/aerolab-logo.svg' alt='Aerolab Logo' />
);

export default Logo;
