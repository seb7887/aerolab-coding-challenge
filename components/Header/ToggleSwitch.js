import React from 'react';
import styled from 'styled-components';

const Switch = styled.a`
  margin-left: 10px;
  background-color: ${props => props.theme.grey};
  border-radius: 30px;
  border: none;
  height: 30px;
  outline: none;
  position: relative;
  width: 60px;

  &::after {
    background-color: ${props => props.theme.orange};
    border-radius: inherit;
    content: '';
    height: inherit;
    cursor: pointer;
    left: 0;
    top: 0;
    position: absolute;
    transition: left 0.6s, width 0.6s;
    width: 30px;
  }

  &:active:after {
    width: 37.5px;
  }

  &.active:active:after {
    left: calc(100% - 37.5px);
  }

  &.active:after {
    background-color: ${props => props.theme.green};
    left: calc(100% - 30px);
  }
`;

const ToggleSwitch = ({ isActive, changeTheme }) => (
  <>
    Dark Theme <Switch className={isActive && 'active'} onClick={changeTheme} />
  </>
);

export default ToggleSwitch;
