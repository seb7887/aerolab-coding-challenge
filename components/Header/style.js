import styled from 'styled-components';

export const Navbar = styled.nav`
  display: flex;
  flex-direction: column;

  .top-container {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
  }

  .bottom-container {
    display: flex;
    flex-direction: row;
    justify-content: flex-end;
    align-items: center;
    margin-top: 0.8rem;
  }
`;

export const PointsContainer = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  cursor: pointer;

  .name {
    margin: 0 7px 0 0;
    color: ${props => props.theme.black};
  }

  .touchable {
    letter-spacing: 1px;
    display: flex;
    justify-content: center;
    align-items: center;
    flex-direction: row;
    text-align: center;
    padding: 8px 8px;
    border-radius: 20px;
    background-color: ${props => props.theme.bg};

    h2 {
      margin: 0 5px 0 0;
    }
  }
`;
