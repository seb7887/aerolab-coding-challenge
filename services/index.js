export { getCurrentUser } from './getCurrentUser';
export { addPoints } from './addPoints';
export { getProducts } from './getProducts';
export { redeem } from './redeem';
export { getHistory } from './getHistory';
