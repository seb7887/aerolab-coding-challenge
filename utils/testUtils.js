import faker from 'faker';

export const fakeUser = {
  name: 'John Kite',
  points: '7000'
};

export const mockCategories = [
  'all',
  'gaming',
  'laptops',
  'tablets',
  'cameras',
  'audio',
  'tv',
  'drones',
  'accessories',
  'home'
];

export const fakeProduct = {
  _id: 7,
  name: 'iPhone',
  category: 'Phones',
  cost: 800,
  img: {
    url: '/static/image/image.png'
  }
};

export const generateProduct = () => {
  const newProduct = {
    _id: faker.random.uuid(),
    name: faker.commerce.productName(),
    category: faker.commerce.productAdjective(),
    cost: 800,
    img: {
      url: faker.image.imageUrl()
    }
  };
  return newProduct;
};

export const fakeProducts = [fakeProduct];

export const generateProducts = quantity => {
  let products = [];

  for (let i = 0; i < quantity; i++) {
    const newProduct = generateProduct();
    products.push(newProduct);
  }

  return products;
};
