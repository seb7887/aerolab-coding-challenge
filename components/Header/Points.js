import React from 'react';

import Coin from '../shared/Coin';

import { PointsContainer } from './style';

const Points = ({ name, points }) => (
  <PointsContainer>
    <h2>{`${name}`}</h2>
    <div>
      <h2>{`${points}`}</h2>
      <Coin />
    </div>
  </PointsContainer>
);

export default Points;
