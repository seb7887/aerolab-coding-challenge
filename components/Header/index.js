import React from 'react';
import Router from 'next/router';
import Link from 'next/link';
import NProgress from 'nprogress';

import { addPoints } from '../../services';

import Logo from './Logo';
import Points from './Points';
import HistoryIcon from './HistoryIcon';
import ToggleSwitch from './ToggleSwitch';

import { Navbar } from './style';

Router.onRouteChangeStart = () => {
  NProgress.start();
};

Router.onRouteChangeComplete = () => {
  NProgress.done();
};

Router.onRouteChangeError = () => {
  NProgress.done();
};

class Header extends React.Component {
  addPoints = async () => {
    await addPoints();
    Router.push('/');
  };

  render() {
    const { me, changeTheme, isActive } = this.props;
    return (
      <Navbar data-testid='header'>
        <div className='top-container'>
          <Link href='/'>
            <a>
              <Logo />
            </a>
          </Link>
          <Points
            name={me.name}
            points={me.points}
            addPoints={this.addPoints}
          />
        </div>
        <hr />
        <div className='bottom-container'>
          <ToggleSwitch isActive={isActive} changeTheme={changeTheme} />
          <Link href='/history'>
            <a>
              <HistoryIcon />
            </a>
          </Link>
        </div>
      </Navbar>
    );
  }
}

export default Header;
