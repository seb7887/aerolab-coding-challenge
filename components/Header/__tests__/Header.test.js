import { render, fireEvent } from 'react-testing-library';
import Router from 'next/router';

import { addPoints as mockAddPointsService } from '../../../services';
import Header from '../index';
import { fakeUser } from '../../../utils/testUtils';

const mockChangeTheme = jest.fn();

jest.mock('../../../services', () => {
  return {
    addPoints: jest.fn(() => Promise.resolve())
  };
});

jest.mock('next/link', () => {
  return ({ children }) => {
    return children;
  };
});

Router.router = {
  pathname: '/',
  push: () => {},
  prefetch: () => {}
};

afterEach(() => {
  mockAddPointsService.mockClear();
});

describe('Header', () => {
  it('renders and matches snapshot', () => {
    const { asFragment } = render(
      <Header me={fakeUser} changeTheme={mockChangeTheme} isActive={false} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should display user name and points', () => {
    const { getByTestId } = render(
      <Header me={fakeUser} changeTheme={mockChangeTheme} isActive={false} />
    );

    expect(getByTestId('header').textContent).toContain(fakeUser.name);
    expect(getByTestId('header').textContent).toContain(fakeUser.points);
  });

  it('should add points', () => {
    const { getByTestId } = render(
      <Header me={fakeUser} changeTheme={mockChangeTheme} isActive={false} />
    );

    const points = getByTestId('points');
    fireEvent.click(points);

    expect(mockAddPointsService).toHaveBeenCalledTimes(1);
  });
});
