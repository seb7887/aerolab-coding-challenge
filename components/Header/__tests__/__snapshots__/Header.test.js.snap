// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders and matches snapshot 1`] = `
<DocumentFragment>
  <nav
    class="style__Navbar-hz5b73-0 cfNmiU"
    data-testid="header"
  >
    <div
      class="top-container"
    >
      <a>
        <img
          alt="Aerolab Logo"
          src="/static/images/aerolab-logo.svg"
        />
      </a>
      <div
        class="style__PointsContainer-hz5b73-1 hPecVr"
        data-testid="points"
      >
        <h2
          class="name"
        >
          John Kite
        </h2>
        <div
          class="touchable"
        >
          <h2>
            7000
          </h2>
          <img
            alt="Coin"
            class="Coin__Img-sc-1i19xvt-0 cnqtJE"
            src="/static/images/coin.svg"
          />
        </div>
      </div>
    </div>
    <hr />
    <div
      class="bottom-container"
    >
      Dark Theme 
      <a
        class="ToggleSwitch__Switch-sc-1oi44tq-0 bYYtGF"
      />
      <a>
        <img
          alt="History"
          src="/static/images/history.svg"
        />
      </a>
    </div>
  </nav>
</DocumentFragment>
`;
